arxiv/document/metadata.py

```diff
@@ -0,0 +1,313 @@
+"""Representations of arXiv document metadata."""
+from collections import abc
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Iterator, List, Optional, Set, Literal
+
+from ..taxonomy import definitions
+from ..taxonomy.category import Category, Group, Archive
+from ..identifier import Identifier
+from ..license import License
+from .version import VersionEntry
+
+from ..base.urls import canonical_url
+
+
+@dataclass(frozen=True)
+class Submitter:
+    """Represents the person who submitted an arXiv article."""
+
+    name: str
+    """Full name."""
+
+    email: str
+    """Email address."""
+
+    __slots__ = ['name', 'email']
+
+
+@dataclass(frozen=True)
+class AuthorList:
+    """Represents author names."""
+
+    raw: str
+    """Raw author field string."""
+
+    __slots__ = ['raw']
+
+    def __str__(self) -> str:
+        """Return the string representation of AuthorList."""
+        return self.raw
+
+
+@dataclass(frozen=True)
+class DocMetadata:
+    """Class for representing the core arXiv document metadata."""
+
+    raw_safe: str
+    """The raw abs string without submitter email address."""
+
+    arxiv_id: str
+    """arXiv paper identifier string without version affix, e.g. 1810.12345."""
+
+    arxiv_id_v: str
+    """arXiv paper identifier string with version affix, e.g. 1810.12345v2."""
+
+    arxiv_identifier: Identifier
+    """arXiv Identifier for this paper."""
+
+    title: str
+    abstract: str
+
+    modified: datetime
+    """Datetime this version was modified."""
+
+    authors: AuthorList
+    """Article authors."""
+
+    submitter: Submitter
+    """Submitter of the article."""
+
+    
+
+    categories: Optional[str]
+    """Article classification (raw string)."""
+
+    primary_category: Optional[Category]
+    """Primary category."""
+
+    primary_archive: Archive
+    primary_group: Group
+
+    secondary_categories: List[Category]
+    """Secondary categor(y|ies)."""
+
+    journal_ref: Optional[str] = None
+    """Report number."""
+
+    report_num: Optional[str] = None
+    """Report number."""
+
+    doi: Optional[str] = None
+    """Digital Object Identifier (DOI)."""
+
+    acm_class: Optional[str] = None
+    """Association for Computing Machinery (ACM) classification(s)."""
+
+    msc_class: Optional[str] = None
+    """American Mathematical Society Mathematics Subject (MSC)
+       classification(s)."""
+
+    proxy: Optional[str] = None
+    """Proxy submitter."""
+
+    comments: Optional[str] = None
+    """Submitter- and/or administrator-provided comments about the article."""
+
+    __source_format_options__ = Literal['tex', 'pdftex', 'ps', 'html', ]
+    source_format: Optional[__source_format_options__] = None
+
+    version: int = 1
+    """Version of this article."""
+
+    license: License = field(default_factory=License)
+    """License associated with the article."""
+
+    version_history: List[VersionEntry] = field(default_factory=list)
```

Comment:
im not sure if we want version history to be a part of all our DocMetadata objects, version history isnt a column in the doc metadata table and would require additional fetching of all version entries for a document when they may not be needed/wanted
Maybe whetever relies on version history could instead take a List of DocMetadata objects?

Reply:
The VersionEntity is outside the ORM but if it were more directly supported we could specify loading techniques: https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html